tests/test_optim/test_scheduler/test_lr_scheduler.py

```diff
@@ -361,11 +366,162 @@ def test_cosine_restart_scheduler(self):
+            with self.assertRaises(AssertionError):
+                ReduceOnPlateauLR(self.optimizer, last_step=last_step)
+
+        with self.assertRaises(ValueError):
+            ReduceOnPlateauLR(self.optimizer, factor=2.0)
+        with self.assertRaises(ValueError):
+            ReduceOnPlateauLR(self.optimizer, min_value=[0.1, 0.1])
+            ReduceOnPlateauLR(self.optimizer, min_value=[0.1, 0.1, 0.1, 0.1])
+        with self.assertRaises(ValueError):
+            ReduceOnPlateauLR(self.optimizer, threshold=-1.0)
+        with self.assertRaises(ValueError):
+            ReduceOnPlateauLR(self.optimizer, rule='foo')
+        with self.assertRaises(ValueError):
+            ReduceOnPlateauLR(self.optimizer, threshold_rule='foo')
+
+        # Test error in step method
+        scheduler = ReduceOnPlateauLR(self.optimizer, monitor='loss')
+        scheduler.step()
+
+        with self.assertRaises(TypeError):
+            scheduler.step(('foo', 1.0))
+
+        metrics = dict(loss_foo=1.0)
+        with self.assertRaises(KeyError):
+            scheduler.step(metrics)
+
+        # Test scheduler value
+        epoch = 10
+        lr = 0.05
+        factor = 0.1
+        patience = 2
+        cooldown = 1
+        metrics = dict(loss=1.0)
+        single_targets = [
+            0.05,  # first
+            0.05,  # in cooldown
+            0.05,  # (num_bad_epochs = 1) < patience
+            0.05,  # (num_bad_epochs = 2) = patience
+            # (num_bad_epochs = 3) > patience
+            # reset num_bad_epochs cooldown
+            0.005,
+            0.005,  # in cooldown
+            0.005,  # (num_bad_epochs = 1) < patience
+            0.005,  # (num_bad_epochs = 2) = patience
+            # (num_bad_epochs = 3) > patience
+            # reset num_bad_epochs cooldown
+            0.0005,
+            0.0005,  # in cooldown
+        ]
+
+        targets = [
+            single_targets, [t * self.layer2_mult for t in single_targets]
+        ]
+
+        for rule in ('less', 'greater'):
+            for threshold_rule in ('rel', 'abs'):
+                scheduler = ReduceOnPlateauLR(
+                    self.optimizer,
+                    monitor='loss',
+                    rule=rule,
+                    factor=factor,
+                    patience=patience,
+                    threshold_rule=threshold_rule,
+                    cooldown=cooldown,
+                    min_value=0.0,
+                )
+                self._test_scheduler_value(
+                    scheduler, targets, epochs=epoch, step_args=(metrics, ))
+
+                self.optimizer = optim.SGD(
+                    [{
+                        'params': self.model.conv1.parameters()
+                    }, {
+                        'params': self.model.conv2.parameters(),
+                        'lr': lr * self.layer2_mult,
+                    }],
+                    lr=lr,
+                    momentum=0.01,
+                    weight_decay=5e-4)
```

Comment:
Suggest testing more complicated metrics(rather than constant), and make sure the `less`, `greater`, `rel`, `abs` modes can work as expected.

Reply:
ok